File: LeetCode/Strings/3614-Process-String-with-Special-Operations-II.js
/**
 * Problem: 3614. Process String with Special Operations II
 * Platform: LeetCode
 * Difficulty: Medium
 *
 * Approach:
 * 1. Compute the final length without constructing the string.
 * 2. Traverse the operations in reverse.
 * 3. Map k back through each operation until the original character is found.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var processStr = function(s, k) {
    let length = 0;

    // Calculate final length
    for (const ch of s) {
        if (ch >= 'a' && ch <= 'z') {
            length++;
        } else if (ch === '*') {
            if (length > 0) {
                length--;
            }
        } else if (ch === '#') {
            length *= 2;
        }
    }

    if (k >= length) {
        return ".";
    }

    // Reverse simulation
    for (let i = s.length - 1; i >= 0; i--) {
        const ch = s[i];

        if (ch === '*') {
            length++;

        } else if (ch === '%') {
            k = length - 1 - k;

        } else if (ch === '#') {
            length = Math.floor(length / 2);

            if (k >= length) {
                k -= length;
            }

        } else {
            length--;

            if (length === k) {
                return ch;
            }
        }
    }

    return ".";
};

File: LeetCode/Strings/3614-Process-String-with-Special-Operations-II.py
"""
Problem: 3614. Process String with Special Operations II
Platform: LeetCode
Difficulty: Medium

Approach:
Simulate the operations using a string builder.

Operations:
- lowercase letter: append
- '*': remove last character
- '#': duplicate current string
- '%': reverse current string

Time Complexity: O(n²) worst case
Space Complexity: O(n)
"""

class Solution:
    def processStr(self, s: str, k: int) -> str:
        length = 0
        for ch in s: 
            if ch>= "a" and ch <= "z":
                length+=1
            elif ch == "*":
                if length > 0:
                    length -=1
            elif ch == "#":
                length *= 2
        if k >= length:
            return "."
        for i in range(len(s)-1, -1, -1):
            ch = s[i]
            if ch == "*":
                length += 1
            elif ch == "%" :
                k = length-1-k
            elif ch == "#":
                length /=2
                if(k>= length):
                    k = k-length
            else:
                length-=1
                if length == k:
                    return ch
        return "."